File: t/VB/MrcSegmentationDemo/UI/SaveMultiPageDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Collections.Generic

Imports Leadtools
Imports Leadtools.Mrc

Namespace MrcSegmentationDemo
   ''' <summary>
   ''' Summary description for SaveMultiPageDialog.
   ''' </summary>
   Public Class SaveMultiPageDialog : Inherits System.Windows.Forms.Form
      Private WithEvents _btnCancel As System.Windows.Forms.Button
      Private WithEvents _btnSave As System.Windows.Forms.Button
      Private _cbSaveFormat As System.Windows.Forms.ComboBox
      Private _lstFilesName As System.Windows.Forms.ListBox
      Private WithEvents _btnUp As System.Windows.Forms.Button
      Private WithEvents _btnDown As System.Windows.Forms.Button
      Private _lblSaveFormat As System.Windows.Forms.Label
      ''' <summary>
      ''' Required designer variable.
      ''' </summary>
      Private components As System.ComponentModel.Container = Nothing

      Public Sub New(ByVal mainForm As MainForm)
         '
         ' Required for Windows Form Designer support
         '
         InitializeComponent()

         '
         ' TODO: Add any constructor code after InitializeComponent call
         '
         _mainForm = mainForm
         InitClass()
      End Sub

      ''' <summary>
      ''' Clean up any resources being used.
      ''' </summary>
      Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
         If disposing Then
            If Not components Is Nothing Then
               components.Dispose()
            End If
         End If
         MyBase.Dispose(disposing)
      End Sub

#Region "Windows Form Designer generated code"
      ''' <summary>
      ''' Required method for Designer support - do not modify
      ''' the contents of this method with the code editor.
      ''' </summary>
      Private Sub InitializeComponent()
         Dim resources As System.Resources.ResourceManager = New System.Resources.ResourceManager(GetType(SaveMultiPageDialog))
         Me._btnCancel = New System.Windows.Forms.Button()
         Me._btnSave = New System.Windows.Forms.Button()
         Me._cbSaveFormat = New System.Windows.Forms.ComboBox()
         Me._lstFilesName = New System.Windows.Forms.ListBox()
         Me._btnUp = New System.Windows.Forms.Button()
         Me._btnDown = New System.Windows.Forms.Button()
         Me._lblSaveFormat = New System.Windows.Forms.Label()
         Me.SuspendLayout()
         ' 
         ' _btnCancel
         ' 
         Me._btnCancel.DialogResult = System.Windows.Forms.DialogResult.Cancel
         Me._btnCancel.FlatStyle = System.Windows.Forms.FlatStyle.System
         Me._btnCancel.Location = New System.Drawing.Point(96, 144)
         Me._btnCancel.Name = "_btnCancel"
         Me._btnCancel.TabIndex = 6
         Me._btnCancel.Text = "&Cancel"
         ' 
         ' _btnSave
         ' 
         Me._btnSave.DialogResult = System.Windows.Forms.DialogResult.OK
         Me._btnSave.FlatStyle = System.Windows.Forms.FlatStyle.System
         Me._btnSave.Location = New System.Drawing.Point(8, 144)
         Me._btnSave.Name = "_btnSave"
         Me._btnSave.TabIndex = 5
         Me._btnSave.Text = "&Save"
         ' 
         ' _cbSaveFormat
         ' 
         Me._cbSaveFormat.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
         Me._cbSaveFormat.Location = New System.Drawing.Point(84, 109)
         Me._cbSaveFormat.Name = "_cbSaveFormat"
         Me._cbSaveFormat.Size = New System.Drawing.Size(100, 21)
         Me._cbSaveFormat.TabIndex = 4
         ' 
         ' _lstFilesName
         ' 
         Me._lstFilesName.HorizontalScrollbar = True
         Me._lstFilesName.Location = New System.Drawing.Point(8, 8)
         Me._lstFilesName.Name = "_lstFilesName"
         Me._lstFilesName.Size = New System.Drawing.Size(128, 95)
         Me._lstFilesName.TabIndex = 0
         ' 
         ' _btnUp
         ' 
         Me._btnUp.BackgroundImage = (CType(resources.GetObject("_btnUp.BackgroundImage"), System.Drawing.Image))
         Me._btnUp.FlatStyle = System.Windows.Forms.FlatStyle.System
         Me._btnUp.Location = New System.Drawing.Point(144, 24)
         Me._btnUp.Name = "_btnUp"
         Me._btnUp.Size = New System.Drawing.Size(48, 24)
         Me._btnUp.TabIndex = 1
         Me._btnUp.Text = "&Up"
         ' 
         ' _btnDown
         ' 
         Me._btnDown.FlatStyle = System.Windows.Forms.FlatStyle.System
         Me._btnDown.Location = New System.Drawing.Point(144, 64)
         Me._btnDown.Name = "_btnDown"
         Me._btnDown.Size = New System.Drawing.Size(48, 24)
         Me._btnDown.TabIndex = 2
         Me._btnDown.Text = "&Down"
         ' 
         ' _lblSaveFormat
         ' 
         Me._lblSaveFormat.Location = New System.Drawing.Point(8, 112)
         Me._lblSaveFormat.Name = "_lblSaveFormat"
         Me._lblSaveFormat.Size = New System.Drawing.Size(72, 16)
         Me._lblSaveFormat.TabIndex = 3
         Me._lblSaveFormat.Text = "Save &Format"
         ' 
         ' SaveMultiPageDialog
         ' 
         Me.AcceptButton = Me._btnSave
         Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
         Me.CancelButton = Me._btnCancel
         Me.ClientSize = New System.Drawing.Size(202, 175)
         Me.Controls.Add(Me._lblSaveFormat)
         Me.Controls.Add(Me._btnDown)
         Me.Controls.Add(Me._btnUp)
         Me.Controls.Add(Me._lstFilesName)
         Me.Controls.Add(Me._cbSaveFormat)
         Me.Controls.Add(Me._btnSave)
         Me.Controls.Add(Me._btnCancel)
         Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog
         Me.MaximizeBox = False
         Me.MinimizeBox = False
         Me.Name = "SaveMultiPageDialog"
         Me.ShowInTaskbar = False
         Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent
         Me.Text = "Save Multi Page"
         Me.ResumeLayout(False)

      End Sub
#End Region

      Private _mainForm As MainForm
      Private _mdiChildren As ViewerForm()
      Private _imageCollection As List(Of RasterImage)

      Private _mrcSegmenterCollection As List(Of MrcSegmenter)

      Public ReadOnly Property SegmenterCollection() As List(Of MrcSegmenter)
         Get
            Return _mrcSegmenterCollection
         End Get
      End Property

      Public ReadOnly Property ImageCollection() As List(Of RasterImage)
         Get
            Return _imageCollection
         End Get
      End Property

      Public ReadOnly Property SaveType() As Integer
         Get
            Return _cbSaveFormat.SelectedIndex
         End Get
      End Property

      Private Sub InitClass()
         _mdiChildren = New ViewerForm(_mainForm.MdiChildren.Length - 1) {}
         Dim index As Integer
         index = 0
         Do While index < _mainForm.MdiChildren.Length
            _mdiChildren(index) = CType(_mainForm.MdiChildren(index), ViewerForm)
            _lstFilesName.Items.Add(_mdiChildren(index).Text)
            index += 1
         Loop

         _cbSaveFormat.Items.Add("Standard Mrc")
         _cbSaveFormat.Items.Add("LEAD Mrc")
         _cbSaveFormat.Items.Add("PDF")
         _cbSaveFormat.SelectedIndex = 0
         _lstFilesName.SelectedIndex = index - 1
         _imageCollection = New List(Of RasterImage)()
         _mrcSegmenterCollection = New List(Of MrcSegmenter)()
      End Sub

      Private Sub _btnUp_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles _btnUp.Click
         Dim currentIndex As Integer = _lstFilesName.SelectedIndex

         If currentIndex = 0 Then
            Return
         End If

         Dim tempString As String = _lstFilesName.Items(_lstFilesName.SelectedIndex).ToString()

         _lstFilesName.Items.RemoveAt(currentIndex)
         _lstFilesName.Items.Insert(currentIndex - 1, tempString)
         _lstFilesName.SelectedIndex = currentIndex - 1

         RearrangeFormsArray(currentIndex, currentIndex - 1)
      End Sub

      Private Sub _btnDown_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles _btnDown.Click
         Dim currentIndex As Integer = _lstFilesName.SelectedIndex

         If currentIndex = _lstFilesName.Items.Count - 1 Then
            Return
         End If

         Dim tempString As String = _lstFilesName.Items(_lstFilesName.SelectedIndex).ToString()

         _lstFilesName.Items.RemoveAt(currentIndex)
         _lstFilesName.Items.Insert(currentIndex + 1, tempString)
         _lstFilesName.SelectedIndex = currentIndex + 1
      End Sub

      Private Sub _btnSave_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles _btnSave.Click
         Dim index As Integer
         index = 0
         Do While index < _lstFilesName.Items.Count
            _imageCollection.Add(_mdiChildren(index).Image)
            If _mdiChildren(index).ViewerSegmenter Is Nothing Then
               _mrcSegmenterCollection.Add(_mdiChildren(index).GetNewSegmenter())
            Else
               _mrcSegmenterCollection.Add(_mdiChildren(index).ViewerSegmenter)
            End If
            index += 1
         Loop
      End Sub

      Private Sub RearrangeFormsArray(ByVal firstIndex As Integer, ByVal secondIndex As Integer)
         Dim tempViwerForm As ViewerForm = _mdiChildren(secondIndex)

         _mdiChildren(secondIndex) = _mdiChildren(firstIndex)

         _mdiChildren(firstIndex) = tempViwerForm
      End Sub

      Private Sub _btnCancel_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles _btnCancel.Click
         Close()
      End Sub
   End Class
End Namespace

File: t/CS/MrcSegmentationDemo/UI/SaveMultiPageDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Collections.Generic;

using Leadtools;
using Leadtools.Mrc;

namespace MrcSegmentationDemo
{
   /// <summary>
   /// Summary description for SaveMultiPageDialog.
   /// </summary>
   public class SaveMultiPageDialog : System.Windows.Forms.Form
   {
      private System.Windows.Forms.Button _btnCancel;
      private System.Windows.Forms.Button _btnSave;
      private System.Windows.Forms.ComboBox _cbSaveFormat;
      private System.Windows.Forms.ListBox _lstFilesName;
      private System.Windows.Forms.Button _btnUp;
      private System.Windows.Forms.Button _btnDown;
      private System.Windows.Forms.Label _lblSaveFormat;
      /// <summary>
      /// Required designer variable.
      /// </summary>
      private System.ComponentModel.Container components = null;

      public SaveMultiPageDialog(MainForm mainForm)
      {
         //
         // Required for Windows Form Designer support
         //
         InitializeComponent();

         //
         // TODO: Add any constructor code after InitializeComponent call
         //
         _mainForm = mainForm;
         InitClass();
      }

      /// <summary>
      /// Clean up any resources being used.
      /// </summary>
      protected override void Dispose(bool disposing)
      {
         if(disposing)
         {
            if(components != null)
            {
               components.Dispose();
            }
         }
         base.Dispose(disposing);
      }

      #region Windows Form Designer generated code
      /// <summary>
      /// Required method for Designer support - do not modify
      /// the contents of this method with the code editor.
      /// </summary>
      private void InitializeComponent( )
      {
         System.Resources.ResourceManager resources = new System.Resources.ResourceManager(typeof(SaveMultiPageDialog));
         this._btnCancel = new System.Windows.Forms.Button();
         this._btnSave = new System.Windows.Forms.Button();
         this._cbSaveFormat = new System.Windows.Forms.ComboBox();
         this._lstFilesName = new System.Windows.Forms.ListBox();
         this._btnUp = new System.Windows.Forms.Button();
         this._btnDown = new System.Windows.Forms.Button();
         this._lblSaveFormat = new System.Windows.Forms.Label();
         this.SuspendLayout();
         // 
         // _btnCancel
         // 
         this._btnCancel.DialogResult = System.Windows.Forms.DialogResult.Cancel;
         this._btnCancel.FlatStyle = System.Windows.Forms.FlatStyle.System;
         this._btnCancel.Location = new System.Drawing.Point(96, 144);
         this._btnCancel.Name = "_btnCancel";
         this._btnCancel.TabIndex = 6;
         this._btnCancel.Text = "&Cancel";
         this._btnCancel.Click += new System.EventHandler(this._btnCancel_Click);
         // 
         // _btnSave
         // 
         this._btnSave.DialogResult = System.Windows.Forms.DialogResult.OK;
         this._btnSave.FlatStyle = System.Windows.Forms.FlatStyle.System;
         this._btnSave.Location = new System.Drawing.Point(8, 144);
         this._btnSave.Name = "_btnSave";
         this._btnSave.TabIndex = 5;
         this._btnSave.Text = "&Save";
         this._btnSave.Click += new System.EventHandler(this._btnSave_Click);
         // 
         // _cbSaveFormat
         // 
         this._cbSaveFormat.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
         this._cbSaveFormat.Location = new System.Drawing.Point(84, 109);
         this._cbSaveFormat.Name = "_cbSaveFormat";
         this._cbSaveFormat.Size = new System.Drawing.Size(100, 21);
         this._cbSaveFormat.TabIndex = 4;
         // 
         // _lstFilesName
         // 
         this._lstFilesName.HorizontalScrollbar = true;
         this._lstFilesName.Location = new System.Drawing.Point(8, 8);
         this._lstFilesName.Name = "_lstFilesName";
         this._lstFilesName.Size = new System.Drawing.Size(128, 95);
         this._lstFilesName.TabIndex = 0;
         // 
         // _btnUp
         // 
         this._btnUp.BackgroundImage = ((System.Drawing.Image)(resources.GetObject("_btnUp.BackgroundImage")));
         this._btnUp.FlatStyle = System.Windows.Forms.FlatStyle.System;
         this._btnUp.Location = new System.Drawing.Point(144, 24);
         this._btnUp.Name = "_btnUp";
         this._btnUp.Size = new System.Drawing.Size(48, 24);
         this._btnUp.TabIndex = 1;
         this._btnUp.Text = "&Up";
         this._btnUp.Click += new System.EventHandler(this._btnUp_Click);
         // 
         // _btnDown
         // 
         this._btnDown.FlatStyle = System.Windows.Forms.FlatStyle.System;
         this._btnDown.Location = new System.Drawing.Point(144, 64);
         this._btnDown.Name = "_btnDown";
         this._btnDown.Size = new System.Drawing.Size(48, 24);
         this._btnDown.TabIndex = 2;
         this._btnDown.Text = "&Down";
         this._btnDown.Click += new System.EventHandler(this._btnDown_Click);
         // 
         // _lblSaveFormat
         // 
         this._lblSaveFormat.Location = new System.Drawing.Point(8, 112);
         this._lblSaveFormat.Name = "_lblSaveFormat";
         this._lblSaveFormat.Size = new System.Drawing.Size(72, 16);
         this._lblSaveFormat.TabIndex = 3;
         this._lblSaveFormat.Text = "Save &Format";
         // 
         // SaveMultiPageDialog
         // 
         this.AcceptButton = this._btnSave;
         this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
         this.CancelButton = this._btnCancel;
         this.ClientSize = new System.Drawing.Size(202, 175);
         this.Controls.Add(this._lblSaveFormat);
         this.Controls.Add(this._btnDown);
         this.Controls.Add(this._btnUp);
         this.Controls.Add(this._lstFilesName);
         this.Controls.Add(this._cbSaveFormat);
         this.Controls.Add(this._btnSave);
         this.Controls.Add(this._btnCancel);
         this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
         this.MaximizeBox = false;
         this.MinimizeBox = false;
         this.Name = "SaveMultiPageDialog";
         this.ShowInTaskbar = false;
         this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
         this.Text = "Save Multi Page";
         this.ResumeLayout(false);

      }
      #endregion

      private MainForm _mainForm;
      private ViewerForm[] _mdiChildren;
      private List<RasterImage> _imageCollection;

      private List<MrcSegmenter> _mrcSegmenterCollection;

      public List<MrcSegmenter> SegmenterCollection
      {
         get
         {
            return _mrcSegmenterCollection;
         }
      }

      public List<RasterImage> ImageCollection
      {
         get
         {
            return _imageCollection;
         }
      }

      public int SaveType
      {
         get
         {
            return _cbSaveFormat.SelectedIndex;
         }
      }

      private void InitClass( )
      {
         _mdiChildren = new ViewerForm[_mainForm.MdiChildren.Length];
         int index;
         for(index = 0; index < _mainForm.MdiChildren.Length; index++)
         {
            _mdiChildren[index] = (ViewerForm)_mainForm.MdiChildren[index];
            _lstFilesName.Items.Add(_mdiChildren[index].Text);
         }

         _cbSaveFormat.Items.Add("Standard Mrc");
         _cbSaveFormat.Items.Add("LEAD Mrc");
         _cbSaveFormat.Items.Add("PDF");
         _cbSaveFormat.SelectedIndex = 0;
         _lstFilesName.SelectedIndex = index - 1;
         _imageCollection = new List<RasterImage>();
         _mrcSegmenterCollection = new List<MrcSegmenter>();
      }

      private void _btnUp_Click(object sender, System.EventArgs e)
      {
         int currentIndex = _lstFilesName.SelectedIndex;

         if(currentIndex == 0)
            return;

         string tempString = _lstFilesName.Items[_lstFilesName.SelectedIndex].ToString();

         _lstFilesName.Items.RemoveAt(currentIndex);
         _lstFilesName.Items.Insert(currentIndex - 1, tempString);
         _lstFilesName.SelectedIndex = currentIndex - 1;

         RearrangeFormsArray(currentIndex, currentIndex - 1);
      }

      private void _btnDown_Click(object sender, System.EventArgs e)
      {
         int currentIndex = _lstFilesName.SelectedIndex;

         if(currentIndex == _lstFilesName.Items.Count - 1)
            return;

         string tempString = _lstFilesName.Items[_lstFilesName.SelectedIndex].ToString();

         _lstFilesName.Items.RemoveAt(currentIndex);
         _lstFilesName.Items.Insert(currentIndex + 1, tempString);
         _lstFilesName.SelectedIndex = currentIndex + 1;
      }

      private void _btnSave_Click(object sender, System.EventArgs e)
      {
         int index;
         for(index = 0; index < _lstFilesName.Items.Count; index++)
         {
            _imageCollection.Add(_mdiChildren[index].Image);
            if(_mdiChildren[index].ViewerSegmenter == null)
               _mrcSegmenterCollection.Add(_mdiChildren[index].GetNewSegmenter());
            else
               _mrcSegmenterCollection.Add(_mdiChildren[index].ViewerSegmenter);
         }
      }

      private void RearrangeFormsArray(int firstIndex, int secondIndex)
      {
         ViewerForm tempViwerForm = _mdiChildren[secondIndex];

         _mdiChildren[secondIndex] = _mdiChildren[firstIndex];

         _mdiChildren[firstIndex] = tempViwerForm;
      }

      private void _btnCancel_Click(object sender, System.EventArgs e)
      {
         Close();
      }
   }
}
